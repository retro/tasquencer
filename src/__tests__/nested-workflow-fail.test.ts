import { Effect } from 'effect';
import { it } from 'vitest';

import { Builder, IdGenerator, Service } from '../index.js';
import { getEnabledTaskNames, makeIdGenerator } from './shared.js';

const subWorkflowDefinition = Builder.workflow()
  .withName('sub')
  .withParentContext()
  .startCondition('start')
  .task('subT1')
  .endCondition('end')
  .connectCondition('start', (to) => to.task('subT1'))
  .connectTask('subT1', (to) => to.condition('end'));

const workflowDefinition = Builder.workflow()
  .withName('activities')
  .startCondition('start')
  .task('t1', Builder.emptyTask().withSplitType('and'))
  .task('t2')
  .compositeTask('t3', (ct) => ct().withSubWorkflow(subWorkflowDefinition))
  .task('t4', Builder.emptyTask().withJoinType('and'))
  .endCondition('end')
  .connectCondition('start', (to) => to.task('t1'))
  .connectTask('t1', (to) => to.task('t2').task('t3'))
  .connectTask('t2', (to) => to.task('t4'))
  .connectTask('t3', (to) => to.task('t4'))
  .connectTask('t4', (to) => to.condition('end'));

it('handles workflow failure in nested workflows (1)', ({ expect }) => {
  const program = Effect.gen(function* () {
    const idGenerator = makeIdGenerator();

    const service = yield* workflowDefinition.build().pipe(
      Effect.flatMap((workflow) => Service.initialize(workflow)),
      Effect.provideService(IdGenerator, idGenerator)
    );

    yield* service.start();
    const state = yield* service.getState();
    expect(state).toMatchSnapshot();
    expect(getEnabledTaskNames(state)).toEqual(new Set(['t1']));

    yield* service.startTask('t1');
    const state2 = yield* service.getState();
    expect(state2).toMatchSnapshot();
    expect(getEnabledTaskNames(state2)).toEqual(new Set(['t2', 't3']));

    yield* service.startTask('t2');
    const state3 = yield* service.getState();
    expect(state3).toMatchSnapshot();
    expect(getEnabledTaskNames(state3)).toEqual(new Set(['t3']));

    yield* service.startTask('t3');
    const state4 = yield* service.getState();
    expect(state4).toMatchSnapshot();
    expect(getEnabledTaskNames(state4)).toEqual(new Set());

    yield* service.initializeWorkItem('t2');
    const { id: subWorkflowId } = yield* service.initializeWorkflow('t3');
    yield* service.startWorkflow(`t3.${subWorkflowId}`);
    yield* service.startTask(`t3.${subWorkflowId}.subT1`);
    const { id: workItemId } = yield* service.initializeWorkItem(
      `t3.${subWorkflowId}.subT1`
    );
    yield* service.startWorkItem(`t3.${subWorkflowId}.subT1.${workItemId}`);
    yield* service.failWorkItem(`t3.${subWorkflowId}.subT1.${workItemId}`);
    const state5 = yield* service.getState();
    expect(state5).toMatchSnapshot();
    expect(getEnabledTaskNames(state5)).toEqual(new Set());
  });

  Effect.runSync(program);
});

it('handles workflow failure in nested workflows (2)', ({ expect }) => {
  const program = Effect.gen(function* () {
    const idGenerator = makeIdGenerator();

    const service = yield* workflowDefinition.build().pipe(
      Effect.flatMap((workflow) => Service.initialize(workflow)),
      Effect.provideService(IdGenerator, idGenerator)
    );

    yield* service.start();
    const state = yield* service.getState();
    expect(state).toMatchSnapshot();
    expect(getEnabledTaskNames(state)).toEqual(new Set(['t1']));

    yield* service.startTask('t1');
    const state2 = yield* service.getState();
    expect(state2).toMatchSnapshot();
    expect(getEnabledTaskNames(state2)).toEqual(new Set(['t2', 't3']));

    yield* service.startTask('t2');
    const state3 = yield* service.getState();
    expect(state3).toMatchSnapshot();
    expect(getEnabledTaskNames(state3)).toEqual(new Set(['t3']));

    yield* service.startTask('t3');
    const state4 = yield* service.getState();
    expect(state4).toMatchSnapshot();
    expect(getEnabledTaskNames(state4)).toEqual(new Set());

    const { id: parentWorkItemId } = yield* service.initializeWorkItem('t2');
    const { id: subWorkflowId } = yield* service.initializeWorkflow('t3');
    yield* service.startWorkflow(`t3.${subWorkflowId}`);
    yield* service.startTask(`t3.${subWorkflowId}.subT1`);
    const { id: workItemId } = yield* service.initializeWorkItem(
      `t3.${subWorkflowId}.subT1`
    );
    yield* service.startWorkItem(`t3.${subWorkflowId}.subT1.${workItemId}`);
    yield* service.startWorkItem(`t2.${parentWorkItemId}`);

    yield* service.failWorkItem(`t2.${parentWorkItemId}`);

    const state5 = yield* service.getState();
    expect(state5).toMatchSnapshot();
    expect(getEnabledTaskNames(state5)).toEqual(new Set());
  });

  Effect.runSync(program);
});
