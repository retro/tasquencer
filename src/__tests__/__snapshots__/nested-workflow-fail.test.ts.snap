// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`handles workflow failure in nested workflows (1) 1`] = `
{
  "conditions": [
    {
      "marking": 1,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t1->t2",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t1->t3",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t2->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t3->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "tasks": [
    {
      "generation": 0,
      "name": "t1",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t2",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t3",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t4",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`handles workflow failure in nested workflows (1) 2`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 1,
      "name": "implicit:t1->t2",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 1,
      "name": "implicit:t1->t3",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t2->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t3->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "t1",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t2",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t3",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t4",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`handles workflow failure in nested workflows (1) 3`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t1->t2",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 1,
      "name": "implicit:t1->t3",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t2->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t3->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "t1",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "t2",
      "state": "started",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t3",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t4",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`handles workflow failure in nested workflows (1) 4`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t1->t2",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t1->t3",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t2->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t3->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "t1",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "t2",
      "state": "started",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "t3",
      "state": "started",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t4",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`handles workflow failure in nested workflows (1) 5`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t1->t2",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t1->t3",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t2->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t3->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "t1",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "t2",
      "state": "canceled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "t3",
      "state": "failed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t4",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "subT1",
      "state": "failed",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "workItems": [
    {
      "id": "workItem-1",
      "payload": null,
      "state": "canceled",
      "taskGeneration": 1,
      "taskName": "t2",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "id": "workItem-2",
      "payload": null,
      "state": "failed",
      "taskGeneration": 1,
      "taskName": "subT1",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "failed",
    },
    {
      "context": undefined,
      "id": "workflow-2",
      "name": "sub",
      "parent": {
        "taskGeneration": 1,
        "taskName": "t3",
        "workflowId": "workflow-1",
        "workflowName": "activities",
      },
      "state": "failed",
    },
  ],
}
`;

exports[`handles workflow failure in nested workflows (1) 6`] = `
{
  "conditions": [
    {
      "marking": 1,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t1->t2",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t1->t3",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t2->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t3->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "tasks": [
    {
      "generation": 0,
      "name": "t1",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t2",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t3",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t4",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`handles workflow failure in nested workflows (1) 7`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 1,
      "name": "implicit:t1->t2",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 1,
      "name": "implicit:t1->t3",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t2->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t3->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "t1",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t2",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t3",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t4",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`handles workflow failure in nested workflows (1) 8`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t1->t2",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 1,
      "name": "implicit:t1->t3",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t2->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t3->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "t1",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "t2",
      "state": "started",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t3",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t4",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`handles workflow failure in nested workflows (1) 9`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t1->t2",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t1->t3",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t2->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t3->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "t1",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "t2",
      "state": "started",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "t3",
      "state": "started",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t4",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`handles workflow failure in nested workflows (1) 10`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t1->t2",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t1->t3",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t2->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "implicit:t3->t4",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "t1",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "t2",
      "state": "failed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "t3",
      "state": "canceled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 0,
      "name": "t4",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "generation": 1,
      "name": "subT1",
      "state": "canceled",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "workItems": [
    {
      "id": "workItem-1",
      "payload": null,
      "state": "failed",
      "taskGeneration": 1,
      "taskName": "t2",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "id": "workItem-2",
      "payload": null,
      "state": "canceled",
      "taskGeneration": 1,
      "taskName": "subT1",
      "workflowId": "workflow-2",
      "workflowName": "sub",
    },
  ],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "failed",
    },
    {
      "context": undefined,
      "id": "workflow-2",
      "name": "sub",
      "parent": {
        "taskGeneration": 1,
        "taskName": "t3",
        "workflowId": "workflow-1",
        "workflowName": "activities",
      },
      "state": "canceled",
    },
  ],
}
`;
